### PR TITLE
HPCC-13566 Encryption in transit for Roxie

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by ghalliday — COMMENTED** [review]
@richardkchapman looks fairly clean.  A few minor comments.

File: roxie/udplib/udptrr.cpp [hunk 1]
```diff
@@ -607,7 +613,14 @@ class CReceiveManager : implements IReceiveManager, public CInterface
                 {
                     unsigned int res;
                     b = bufferManager->allocate();
-                    receive_socket->read(b->data, 1, DATA_PAYLOAD, res, 5);
+                    if (parent.encrypted)
+                    {
+                        MemoryBuffer indata;
+                        receive_socket->read(indata.reserveTruncate(DATA_PAYLOAD+50), 1, DATA_PAYLOAD+50, res, 5);
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply roxie/udplib/udptrr.cpp:619]: It's really "round up to next multiple of 16" but I was too lazy to code it at the time!

I will fix

File: roxie/udplib/udptrr.cpp [hunk 2]
```diff
@@ -607,7 +613,14 @@ class CReceiveManager : implements IReceiveManager, public CInterface
                 {
                     unsigned int res;
                     b = bufferManager->allocate();
-                    receive_socket->read(b->data, 1, DATA_PAYLOAD, res, 5);
+                    if (parent.encrypted)
+                    {
+                        MemoryBuffer indata;
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply roxie/udplib/udptrr.cpp:618]: Yes, I thought that at the time and intended to gome back and change it (but forgot).

I will fix

File: roxie/ccd/ccdqueue.cpp [hunk 3]
```diff
@@ -601,11 +621,136 @@ class CRoxieQueryPacket : implements IRoxieQueryPacket, public CInterface
...
+    {
+        unsigned length = data->packetlength;
+        RoxiePacketHeader *newdata = (RoxiePacketHeader *) malloc(length);
+        memcpy(newdata, data, length);
+        newdata->channel = channel;
+        newdata->retries |= ROXIE_BROADCAST;
+        return new CNocryptRoxieQueryPacket(newdata, length);
+    }
+
+    virtual ISerializedRoxieQueryPacket *serialize() const override
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply roxie/ccd/ccdqueue.cpp:682]: I think I tried that and quickly disappeared down a rabbit hole of needing to add consts all over the place, and backed out.

File: roxie/ccd/ccdfile.cpp [hunk 4]
```diff
@@ -755,6 +755,7 @@ class CacheReportingBuffer : public CInterfaceOf<ICacheInfoRecorder>
 private:
     static StringBuffer &appendRange(StringBuffer &ret, offset_t start, offset_t end, bool diskCache)
     {
+        ret.append(' ');
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply roxie/ccd/ccdfile.cpp:758]: Should disappear when I rebase

File: system/jlib/jencrypt.cpp [hunk 5]
```diff
@@ -1798,15 +1805,28 @@ MemoryBuffer &aesDecrypt(const void *key, size_t keylen, const void *input, size
     
     rin.init(Rijndael::CBC, Rijndael::Decrypt, (const UINT8 *)key, keyType);
     size32_t truncInLen = (size32_t)inlen;
-    int len = rin.padDecrypt((const UINT8 *)input, truncInLen, (UINT8 *) output.clear().reserveTruncate(truncInLen));
-
+    size32_t origLen = output.length();
+    int len = rin.padDecrypt((const UINT8 *)input, truncInLen, (UINT8 *) output.reserveTruncate(truncInLen), truncInLen);
```
Comment (mckellyln) [inline system/jlib/jencrypt.cpp:1809]: len will always be <= truncInLen, right ?

File: roxie/ccd/ccdqueue.cpp [hunk 6]
```diff
@@ -622,6 +767,38 @@ extern IRoxieQueryPacket *createRoxiePacket(MemoryBuffer &m)
...
+    // Direct decryption of special packets - others are only decrypted after being dequeued
+    if (encryptInTransit)
+    {
+        RoxiePacketHeader *header = (RoxiePacketHeader *) m.toByteArray();
+        assertex(header != nullptr);
+        assertex(header->activityId == ROXIE_FILECALLBACK || header->activityId == ROXIE_DEBUGCALLBACK);
+        assertex(m.length() >= header->packetlength);
+        unsigned encryptedLen = header->packetlength - sizeof(RoxiePacketHeader);
+        const void *encryptedData = (const void *)(header+1);
+        MemoryBuffer decrypted;
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply roxie/ccd/ccdqueue.cpp:781]: Don't know the final size exactly - could be up to 16 bytes smaller than encryptedLen. Probably still worth it I guess

**Comment by mckellyln** [comment]
It looks good to me.

File: roxie/udplib/udptrr.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ghalliday) [reply roxie/udplib/udptrr.cpp:626]: I don't think you have...

**Review by ghalliday — COMMENTED** [review]
Needs squashing, and one item you said you would change which you haven't

File: roxie/udplib/udptrr.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (richardkchapman) [reply roxie/udplib/udptrr.cpp]: I did, I just forgot to delete the local var. Have removed and squashed

**Comment by richardkchapman** [comment]
@ghalliday Please recheck/merge

**Review by ghalliday — APPROVED** [review]

File: roxie/ccd/ccdqueue.cpp [hunk 7]
```diff
@@ -490,6 +482,49 @@ class CRoxieQueryPacket : implements IRoxieQueryPacket, public CInterface
...
+    }
+
+    ~CRoxieQueryPacketBase()
+    {
+        free(data);
+    }
+
+};
+
+// MORE - this is for TESTING ONLY - do not release with this key here like this!
```
Comment (dcamper) [inline roxie/ccd/ccdqueue.cpp:495]: Does 'release' exclude 'merge into master branch'?

Comment (richardkchapman) [reply roxie/ccd/ccdqueue.cpp:495]: There is another Jira open to fix it properly, but this change allows us to do some performance testing of this option in master. Merging it to master is acceptable so long as that other Jira gets merged before 8.0

**Pull request merged**